lib/elixir/lib/integer.ex

```diff
@@ -240,45 +240,30 @@ defmodule Integer do
   end
 
   def parse(<<?-, rest::binary>>, base) do
-    case parse_digits(rest, base) do
-      {acc, bin} ->
-        {-acc, bin}
-      :error ->
-        :error
-    end
-  end
-
-  def parse(<<?+, rest::binary>>, base) do
-    parse_digits(rest, base)
+    with {int, rem} <- parse_digits(rest, base), do: {-int, rem}
   end
-
-  def parse(<<rest::binary>>, base) do
-    parse_digits(rest, base)
-  end
-
-  digits = [{?0..?9, -?0}, {?A..?Z, 10 - ?A}, {?a..?z, 10 - ?a}]
-
-  for {chars, diff} <- digits, char <- chars do
-    defp parse_digits(<<unquote(char), rest::binary>>, base)
-         when base > unquote(char + diff) do
-      parse_digits(rest, base, unquote(char + diff))
+  def parse(<<?+, rest::binary>>, base), do: parse_digits(rest, base)
+  def parse(<<rest::binary>>, base), do: parse_digits(rest, base)
+
+  defp parse_digits(<<rest::binary>>, base) do
+    case count_digits(rest, base, 0) do
+      0 -> :error
+      n ->
+        {digits, rem} = :erlang.split_binary(rest, n)
```

Comment:
The way we pass the binary around loses the single match context optimisation in this function because `rest` is used twice (see http://erlang.org/doc/efficiency_guide/binaryhandling.html#id65492). If we delay the handling of `+`/`-` to `count_digits` we can use the single match context for the full duration of the parsing. We can avoid the `-` appearing string bug because the `count` argument is matching on `0`, so must appear at the beginning of the string.

It should be possible to add the above clause (and a matching one for `-`), to `count_digits` and only do the binary matching/parsing in the `count_digits` function.